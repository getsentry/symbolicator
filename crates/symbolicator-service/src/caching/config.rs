use std::fmt;

/// All known cache names.
#[derive(Debug, Clone, Copy)]
pub enum CacheName {
    Objects,
    ObjectMeta,
    Auxdifs,
    Il2cpp,
    Symcaches,
    Cficaches,
    PpdbCaches,
    ArtifactCaches,
    SourceMapCaches,
    Diagnostics,
}

impl AsRef<str> for CacheName {
    fn as_ref(&self) -> &str {
        match self {
            Self::Objects => "objects",
            Self::ObjectMeta => "object_meta",
            Self::Auxdifs => "auxdifs",
            Self::Il2cpp => "il2cpp",
            Self::Symcaches => "symcaches",
            Self::Cficaches => "cficaches",
            Self::PpdbCaches => "ppdb_caches",
            Self::ArtifactCaches => "artifact_caches",
            Self::SourceMapCaches => "sourcesmap_caches",
            Self::Diagnostics => "diagnostics",
        }
    }
}

impl fmt::Display for CacheName {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.as_ref())
    }
}
