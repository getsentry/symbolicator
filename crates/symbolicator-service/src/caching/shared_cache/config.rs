use std::path::PathBuf;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilesystemSharedCacheConfig {
    pub path: PathBuf,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GcsSharedCacheConfig {
    /// Name of the GCS bucket.
    pub bucket: String,

    /// Optional name of a JSON file containing the service account credentials.
    ///
    /// If this is not provided the JSON will be looked up in the
    /// `GOOGLE_APPLICATION_CREDENTIALS` variable.  Otherwise it is assumed the service is
    /// running since GCP and will retrieve the correct user or service account from the
    /// GCP services.
    #[serde(default)]
    pub service_account_path: Option<PathBuf>,
}

/// The backend to use for the shared cache.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SharedCacheBackendConfig {
    Gcs(GcsSharedCacheConfig),
    Filesystem(FilesystemSharedCacheConfig),
}

/// A remote cache that can be shared between symbolicator instances.
///
/// Any files not in the local cache will be looked up from here before being looked up in
/// their original source.  Additionally derived caches are also stored in here to save
/// computations if another symbolicator has already done the computation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SharedCacheConfig {
    /// The number of allowed concurrent uploads to the shared cache.
    ///
    /// Uploading to the shared cache is not critical for symbolicator's operation and
    /// should not disrupt any normal work it does.  This limits the number of concurrent
    /// uploads so that associated resources are kept in check.
    #[serde(default = "default_max_concurrent_uploads")]
    pub max_concurrent_uploads: usize,

    /// The number of queued up uploads to the cache.
    ///
    /// If more items need to be uploaded to the shared cache than there are allowed
    /// concurrently the uploads will be queued.  If the queue is full the uploads are
    /// simply dropped as they are not critical to symbolicator's operation and not
    /// disrupting symbolicator is more important than uploading to the shared cache.
    #[serde(default = "default_max_upload_queue_size")]
    pub max_upload_queue_size: usize,

    /// The backend to use for the shared cache.
    #[serde(flatten)]
    pub backend: SharedCacheBackendConfig,
}

fn default_max_upload_queue_size() -> usize {
    400
}

fn default_max_concurrent_uploads() -> usize {
    20
}
