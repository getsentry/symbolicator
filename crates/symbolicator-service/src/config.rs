use std::collections::BTreeMap;
use std::env;
use std::fmt;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::Duration;

use anyhow::{Context, Result};
use sentry::types::Dsn;
use serde::{de, Deserialize, Deserializer};
use tracing::level_filters::LevelFilter;

use symbolicator_sources::SourceConfig;

use crate::cache::SharedCacheConfig;

/// Controls the log format
#[derive(Clone, Copy, Debug, Eq, PartialEq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LogFormat {
    /// Auto detect (pretty for tty, simplified for other)
    Auto,
    /// With colors
    Pretty,
    /// Simplified log output
    Simplified,
    /// Dump out JSON lines
    Json,
}

#[cfg(feature = "https")]
#[derive(Clone, Debug, Eq, PartialEq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub struct HTTPSConfig {
    pub certificate_path: PathBuf,
    pub key_path: PathBuf,
}

#[cfg(feature = "https")]
impl Default for HTTPSConfig {
    fn default() -> Self {
        HTTPSConfig {
            certificate_path: PathBuf::from("cert.pem"),
            key_path: PathBuf::from("cert.pem"),
        }
    }
}

/// Controls the HTTPS server setup
#[cfg(feature = "https")]
#[derive(Clone, Debug, Default, Deserialize)]
#[serde(default)]
pub struct ServerConfig {
    /// HTTPS configuration
    pub https: Option<HTTPSConfig>,
}

/// Controls the logging system.
#[derive(Clone, Debug, Deserialize)]
#[serde(default)]
pub struct Logging {
    /// The log level for the relay.
    #[serde(deserialize_with = "deserialize_level_filter")]
    pub level: LevelFilter,
    /// Controls the log format.
    pub format: LogFormat,
    /// When set to true, backtraces are forced on.
    pub enable_backtraces: bool,
}

impl Default for Logging {
    fn default() -> Self {
        Logging {
            level: LevelFilter::INFO,
            format: LogFormat::Auto,
            enable_backtraces: true,
        }
    }
}

/// Control the metrics.
#[derive(Clone, Debug, Deserialize)]
#[serde(default)]
pub struct Metrics {
    /// host/port of statsd instance
    pub statsd: Option<String>,
    /// The prefix that should be added to all metrics.
    pub prefix: String,
    /// A tag name to report the hostname to, for each metric. Defaults to not sending such a tag.
    pub hostname_tag: Option<String>,
    /// A tag name to report the environment to, for each metric. Defaults to not sending such a tag.
    pub environment_tag: Option<String>,
    /// A map containing custom tags and their values.
    ///
    /// These tags will be appended to every metric.
    pub custom_tags: BTreeMap<String, String>,
}

impl Default for Metrics {
    fn default() -> Self {
        Metrics {
            statsd: match env::var("STATSD_SERVER") {
                Ok(metrics_statsd) => Some(metrics_statsd),
                Err(_) => None,
            },
            prefix: "symbolicator".into(),
            hostname_tag: None,
            environment_tag: None,
            custom_tags: BTreeMap::new(),
        }
    }
}

/// Fine-tuning downloaded cache expiry.
///
/// These differ from [`DerivedCacheConfig`] in the [`Default`] implementation.
#[derive(Debug, Clone, Copy, Deserialize, Eq, PartialEq)]
#[serde(default)]
pub struct DownloadedCacheConfig {
    /// Maximum duration since last use of cache item (item last used).
    #[serde(with = "humantime_serde")]
    pub max_unused_for: Option<Duration>,

    /// Maximum duration since creation of negative cache item (item age).
    #[serde(with = "humantime_serde")]
    pub retry_misses_after: Option<Duration>,

    /// Maximum duration since creation of malformed cache item (item age).
    #[serde(with = "humantime_serde")]
    pub retry_malformed_after: Option<Duration>,

    /// Maximum number of lazy re-downloads
    pub max_lazy_redownloads: isize,
}

impl Default for DownloadedCacheConfig {
    fn default() -> Self {
        Self {
            max_unused_for: Some(Duration::from_secs(3600 * 24)),
            retry_misses_after: Some(Duration::from_secs(3600)),
            retry_malformed_after: Some(Duration::from_secs(3600 * 24)),
            max_lazy_redownloads: 50,
        }
    }
}

/// Fine-tuning derived cache expiry.
///
/// These differ from [`DownloadedCacheConfig`] in the [`Default`] implementation.
#[derive(Debug, Clone, Copy, Deserialize, Eq, PartialEq)]
#[serde(default)]
pub struct DerivedCacheConfig {
    /// Maximum duration since last use of cache item (item last used).
    #[serde(with = "humantime_serde")]
    pub max_unused_for: Option<Duration>,

    /// Maximum duration since creation of negative cache item (item age).
    #[serde(with = "humantime_serde")]
    pub retry_misses_after: Option<Duration>,

    /// Maximum duration since creation of malformed cache item (item age).
    #[serde(with = "humantime_serde")]
    pub retry_malformed_after: Option<Duration>,

    /// Maximum number of lazy re-computations
    pub max_lazy_recomputations: isize,
}

impl Default for DerivedCacheConfig {
    fn default() -> Self {
        Self {
            max_unused_for: Some(Duration::from_secs(3600 * 24 * 7)),
            retry_misses_after: Some(Duration::from_secs(3600)),
            retry_malformed_after: Some(Duration::from_secs(3600 * 24)),
            max_lazy_recomputations: 20,
        }
    }
}

/// Fine-tuning diagnostics caches.
#[derive(Debug, Clone, Copy, Deserialize, Eq, PartialEq)]
#[serde(default)]
pub struct DiagnosticsCacheConfig {
    /// Time to keep diagnostics files cached.
    #[serde(with = "humantime_serde")]
    pub retention: Option<Duration>,
}

impl Default for DiagnosticsCacheConfig {
    fn default() -> Self {
        Self {
            retention: Some(Duration::from_secs(3600 * 24)),
        }
    }
}

/// Struct to treat all cache configs identical in cache code.
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum CacheConfig {
    Downloaded(DownloadedCacheConfig),
    Derived(DerivedCacheConfig),
    Diagnostics(DiagnosticsCacheConfig),
}

impl CacheConfig {
    pub fn max_unused_for(&self) -> Option<Duration> {
        match self {
            Self::Downloaded(cfg) => cfg.max_unused_for,
            Self::Derived(cfg) => cfg.max_unused_for,
            Self::Diagnostics(cfg) => cfg.retention,
        }
    }

    pub fn retry_misses_after(&self) -> Option<Duration> {
        match self {
            Self::Downloaded(cfg) => cfg.retry_misses_after,
            Self::Derived(cfg) => cfg.retry_misses_after,
            Self::Diagnostics(_cfg) => None,
        }
    }

    pub fn retry_malformed_after(&self) -> Option<Duration> {
        match self {
            Self::Downloaded(cfg) => cfg.retry_malformed_after,
            Self::Derived(cfg) => cfg.retry_malformed_after,
            Self::Diagnostics(_cfg) => None,
        }
    }
}

impl From<DownloadedCacheConfig> for CacheConfig {
    fn from(source: DownloadedCacheConfig) -> Self {
        Self::Downloaded(source)
    }
}

impl From<DerivedCacheConfig> for CacheConfig {
    fn from(source: DerivedCacheConfig) -> Self {
        Self::Derived(source)
    }
}

impl From<DiagnosticsCacheConfig> for CacheConfig {
    fn from(source: DiagnosticsCacheConfig) -> Self {
        Self::Diagnostics(source)
    }
}

#[derive(Clone, Debug, Deserialize, Default)]
#[serde(default)]
pub struct CacheConfigs {
    /// Configure how long downloads are cached for.
    pub downloaded: DownloadedCacheConfig,
    /// Configure how long caches derived from downloads are cached for.
    pub derived: DerivedCacheConfig,
    /// How long diagnostics data is cached.
    ///
    /// E.g. minidumps which caused a crash in symbolicator will be stored here.
    pub diagnostics: DiagnosticsCacheConfig,
}

/// See docs/index.md for more information on config values.
#[derive(Clone, Debug, Deserialize)]
#[serde(default)]
pub struct Config {
    /// Which directory to use when caching. Default is not to cache.
    pub cache_dir: Option<PathBuf>,

    /// Host and port to bind the HTTP webserver to.
    pub bind: String,

    /// Host and port to bind the HTTPS webserver to.
    #[cfg(feature = "https")]
    pub bind_https: Option<String>,

    /// Configuration for internal logging.
    pub logging: Logging,

    /// Configuration for reporting metrics to a statsd instance.
    pub metrics: Metrics,

    #[cfg(feature = "https")]
    pub server_config: ServerConfig,

    /// DSN to report internal errors to
    pub sentry_dsn: Option<Dsn>,

    /// Fine-tune cache expiry
    pub caches: CacheConfigs,

    /// Enables symbol proxy mode.
    pub symstore_proxy: bool,

    /// Default list of sources and the sources used for proxy mode.
    pub sources: Arc<[SourceConfig]>,

    /// Allow reserved IP addresses for requests to sources.
    pub connect_to_reserved_ips: bool,

    /// The maximum timeout for downloads.
    ///
    /// This is the upper limit the download service will take for downloading from a single
    /// source, regardless of how many retries are involved. The default is set to 315s,
    /// just above the amount of time it would take for a 4MB/s connection to download 2GB.
    #[serde(with = "humantime_serde")]
    pub max_download_timeout: Duration,

    /// The timeout for the initial HEAD request in a download.
    ///
    /// This timeout applies to each individual attempt to establish a
    /// connection with a symbol source if retries take place.
    #[serde(with = "humantime_serde")]
    pub connect_timeout: Duration,

    /// The timeout per GB for streaming downloads.
    ///
    /// For downloads with a known size, this timeout applies per individual
    /// download attempt. If the download size is not known, it is ignored and
    /// only `max_download_timeout` applies. The default is set to 250s,
    /// just above the amount of time it would take for a 4MB/s connection to
    /// download 1GB.
    #[serde(with = "humantime_serde")]
    pub streaming_timeout: Duration,

    /// The maximum number of requests that symbolicator will process concurrently.
    ///
    /// A value of `None` indicates no limit.
    pub max_concurrent_requests: Option<usize>,

    /// An optional shared cache between multiple symbolicators.
    ///
    /// If configured this cache location is queried whenever a cache item is not found in
    /// the corresponding local cache.  Only if the shared cache does not have the item will
    /// it be locally recreated after which it will be submitted to the shared cache.
    ///
    /// The aim is to make it easy to start up a new symbolicator which will quickly fill up
    /// caches from already running symbolicators.
    pub shared_cache: Option<SharedCacheConfig>,

    /// Internal. Enables crash handling and sets the absolute path to where minidumps should be
    /// cached on disk. The path is created if it doesn't exist. Path must be UTF-8.
    #[serde(default)]
    pub _crash_db: Option<PathBuf>,
}

impl Config {
    /// Return a cache directory `dir`, it is joined with the configured base cache directory.
    ///
    /// If there is no base cache directory configured this means no caching should happen
    /// and this returns None.
    pub fn cache_dir<P>(&self, dir: P) -> Option<PathBuf>
    where
        P: AsRef<Path>,
    {
        self.cache_dir.as_ref().map(|base| base.join(dir))
    }

    pub fn default_sources(&self) -> Arc<[SourceConfig]> {
        self.sources.clone()
    }
}

/// Checks if we are running in docker.
fn is_docker() -> bool {
    if fs::metadata("/.dockerenv").is_ok() {
        return true;
    }

    fs::read_to_string("/proc/self/cgroup")
        .map(|s| s.contains("/docker"))
        .unwrap_or(false)
}

/// Default value for the "bind" configuration.
fn default_bind() -> String {
    if is_docker() {
        // Docker images rely on this service being exposed
        "0.0.0.0:3021".to_owned()
    } else {
        "127.0.0.1:3021".to_owned()
    }
}

/// Default value for the "cache_dir" configuration.
fn default_cache_dir() -> Option<PathBuf> {
    if is_docker() {
        // Docker image already defines `/data` as a persistent volume
        Some(PathBuf::from("/data"))
    } else {
        None
    }
}

impl Default for Config {
    fn default() -> Self {
        Config {
            cache_dir: default_cache_dir(),
            bind: default_bind(),
            #[cfg(feature = "https")]
            bind_https: None,
            logging: Logging::default(),
            #[cfg(feature = "https")]
            server_config: ServerConfig::default(),
            metrics: Metrics::default(),
            sentry_dsn: None,
            caches: CacheConfigs::default(),
            symstore_proxy: true,
            sources: Arc::from(vec![]),
            connect_to_reserved_ips: false,
            // Allow a 4MB/s connection to download 2GB without timing out
            max_download_timeout: Duration::from_secs(315),
            connect_timeout: Duration::from_secs(15),
            // Allow a 4MB/s connection to download 1GB without timing out
            streaming_timeout: Duration::from_secs(250),
            max_concurrent_requests: Some(120),
            shared_cache: None,
            _crash_db: None,
        }
    }
}

impl Config {
    pub fn get(path: Option<&Path>) -> Result<Self> {
        match path {
            Some(path) => Self::from_reader(
                fs::File::open(path).context("failed to open configuration file")?,
            ),
            None => Ok(Config::default()),
        }
    }

    fn from_reader(reader: impl std::io::Read) -> Result<Self> {
        serde_yaml::from_reader(reader).context("failed to parse YAML")
    }
}

#[derive(Debug)]
struct LevelFilterVisitor;

impl<'de> de::Visitor<'de> for LevelFilterVisitor {
    type Value = LevelFilter;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> std::fmt::Result {
        write!(
            formatter,
            r#"one of the strings "off", "error", "warn", "info", "debug", or "trace""#
        )
    }

    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        match v {
            "off" => Ok(LevelFilter::OFF),
            "error" => Ok(LevelFilter::ERROR),
            "warn" => Ok(LevelFilter::WARN),
            "info" => Ok(LevelFilter::INFO),
            "debug" => Ok(LevelFilter::DEBUG),
            "trace" => Ok(LevelFilter::TRACE),
            _ => Err(de::Error::unknown_variant(
                v,
                &["off", "error", "warn", "info", "debug", "trace"],
            )),
        }
    }
}

fn deserialize_level_filter<'de, D: Deserializer<'de>>(
    deserializer: D,
) -> Result<LevelFilter, D::Error> {
    deserializer.deserialize_str(LevelFilterVisitor)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cache_config() {
        // It should be possible to set individual caches in reasonable units without
        // affecting other caches' default values.
        let cfg = Config::get(None).unwrap();
        assert_eq!(
            cfg.caches.diagnostics.retention,
            Some(Duration::from_secs(3600 * 24))
        );

        let yaml = r#"
            caches:
              diagnostics:
                retention: 1h
        "#;
        let cfg = Config::from_reader(yaml.as_bytes()).unwrap();
        assert_eq!(
            cfg.caches.diagnostics.retention,
            Some(Duration::from_secs(3600))
        );

        assert_eq!(cfg.caches.downloaded, DownloadedCacheConfig::default());
        assert_eq!(cfg.caches.derived, DerivedCacheConfig::default());

        let yaml = r#"
            caches:
              downloaded:
                max_unused_for: 500s
        "#;
        let cfg = Config::from_reader(yaml.as_bytes()).unwrap();
        assert_eq!(
            cfg.caches.downloaded.max_unused_for,
            Some(Duration::from_secs(500))
        );
        assert_eq!(
            cfg.caches.downloaded.retry_misses_after,
            Some(Duration::from_secs(3600))
        );
        assert_eq!(
            cfg.caches.downloaded.retry_malformed_after,
            Some(Duration::from_secs(3600 * 24))
        );
        assert_eq!(cfg.caches.derived, DerivedCacheConfig::default());
        assert_eq!(cfg.caches.diagnostics, DiagnosticsCacheConfig::default());
    }

    #[test]
    fn test_disabling_expiry() {
        // It should be possible to set a cache value to `None` meaning "do not expire".
        let yaml = r#"
            caches:
              downloaded:
                max_unused_for: null
        "#;
        let cfg = Config::from_reader(yaml.as_bytes()).unwrap();
        let downloaded_default = DownloadedCacheConfig::default();
        assert_eq!(cfg.caches.downloaded.max_unused_for, None);
        assert_eq!(
            cfg.caches.downloaded.retry_misses_after,
            downloaded_default.retry_misses_after
        );
        assert_eq!(
            cfg.caches.downloaded.retry_malformed_after,
            downloaded_default.retry_malformed_after,
        )
    }

    #[test]
    fn test_unspecified_dl_timeouts() {
        let yaml = r#"
            sources: []
        "#;
        let cfg = Config::from_reader(yaml.as_bytes()).unwrap();
        let default_cfg = Config::default();
        assert_eq!(cfg.max_download_timeout, default_cfg.max_download_timeout);
        assert_eq!(cfg.connect_timeout, default_cfg.connect_timeout);
        assert_eq!(cfg.streaming_timeout, default_cfg.streaming_timeout);
    }

    #[test]
    fn test_zero_second_dl_timeouts() {
        // 0s download timeouts will not be set to defaults
        let yaml = r#"
            max_download_timeout: 0s
            connect_timeout: 0s
            streaming_timeout: 0s
        "#;
        let cfg = Config::from_reader(yaml.as_bytes()).unwrap();
        assert_eq!(cfg.max_download_timeout, Duration::from_secs(0));
        assert_eq!(cfg.connect_timeout, Duration::from_secs(0));
        assert_eq!(cfg.streaming_timeout, Duration::from_secs(0));
    }

    #[test]
    fn test_unknown_fields() {
        // Unknown fields should not cause failure
        let yaml = r#"
            caches:
              not_a_cache:
                max_unused_for: 1h
        "#;
        let cfg = Config::from_reader(yaml.as_bytes());
        assert!(cfg.is_ok());
    }

    #[test]
    fn test_empty_file() {
        // Empty files aren't supported
        let yaml = r#""#;
        let result = Config::from_reader(yaml.as_bytes());
        assert!(result.is_err());
    }
}
