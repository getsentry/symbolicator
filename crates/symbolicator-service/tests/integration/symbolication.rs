use std::sync::Arc;

use symbolicator_service::config::Config;
use symbolicator_service::services::create_service;
use symbolicator_service::services::symbolication::{
    StacktraceOrigin, SymbolicateStacktraces, SymbolicationActor,
};
use symbolicator_service::types::{
    CompleteObjectInfo, RawFrame, RawObjectInfo, RawStacktrace, Scope,
};
use symbolicator_service::utils::hex::HexValue;
use symbolicator_sources::{ObjectType, SourceConfig};
use symbolicator_test as test;
use symbolicator_test::{assert_snapshot, fixture};

/// Setup tests and create a test service.
///
/// This function returns a tuple containing the service to test, and a temporary cache
/// directory. The directory is cleaned up when the [`TempDir`] instance is dropped. Keep it as
/// guard until the test has finished.
///
/// The service is configured with `connect_to_reserved_ips = True`. This allows to use a local
/// symbol server to test object file downloads.
pub fn setup_service(
    update_config: impl FnOnce(&mut Config),
) -> (SymbolicationActor, test::TempDir) {
    test::setup();

    let cache_dir = test::tempdir();

    let mut config = Config {
        cache_dir: Some(cache_dir.path().to_owned()),
        connect_to_reserved_ips: true,
        ..Default::default()
    };
    update_config(&mut config);

    let handle = tokio::runtime::Handle::current();
    let (symbolication, _objects) = create_service(&config, handle).unwrap();

    (symbolication, cache_dir)
}

pub fn get_symbolication_request(sources: Vec<SourceConfig>) -> SymbolicateStacktraces {
    SymbolicateStacktraces {
        scope: Scope::Global,
        signal: None,
        sources: Arc::from(sources),
        origin: StacktraceOrigin::Symbolicate,
        stacktraces: vec![RawStacktrace {
            frames: vec![RawFrame {
                instruction_addr: HexValue(0x1_0000_0fa0),
                ..RawFrame::default()
            }],
            ..RawStacktrace::default()
        }],
        modules: vec![CompleteObjectInfo::from(RawObjectInfo {
            ty: ObjectType::Macho,
            code_id: Some("502fc0a51ec13e479998684fa139dca7".to_owned().to_lowercase()),
            debug_id: Some("502fc0a5-1ec1-3e47-9998-684fa139dca7".to_owned()),
            image_addr: HexValue(0x1_0000_0000),
            image_size: Some(4096),
            code_file: None,
            debug_file: None,
            debug_checksum: None,
        })],
    }
}

pub fn make_symbolication_request(
    modules: Vec<RawObjectInfo>,
    stacktraces: Vec<RawStacktrace>,
    sources: Vec<SourceConfig>,
) -> SymbolicateStacktraces {
    SymbolicateStacktraces {
        modules: modules.into_iter().map(From::from).collect(),
        stacktraces,
        signal: None,
        origin: StacktraceOrigin::Symbolicate,
        sources: Arc::from(sources),
        scope: Default::default(),
    }
}

#[tokio::test]
async fn test_remove_bucket() {
    // Test with sources first, and then without. This test should verify that we do not leak
    // cached debug files to requests that no longer specify a source.

    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_symsrv, source) = test::symbol_server();

    let request = get_symbolication_request(vec![source]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());

    let request = get_symbolication_request(vec![]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_add_bucket() {
    // Test without sources first, then with. This test should verify that we apply a new source
    // to requests immediately.

    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_symsrv, source) = test::symbol_server();

    let request = get_symbolication_request(vec![]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());

    let request = get_symbolication_request(vec![source]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_apple_crash_report() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_symsrv, source) = test::symbol_server();

    let report_file = std::fs::File::open(fixture("apple_crash_report.txt")).unwrap();

    let response = symbolication
        .process_apple_crash_report(Scope::Global, report_file, Arc::new([source]))
        .await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_wasm_payload() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_symsrv, source) = test::symbol_server();

    let modules = serde_json::from_str(
        r#"[{
          "type":"wasm",
          "debug_id":"bda18fd8-5d4a-4eb8-9302-2d6bfad846b1",
          "code_id":"bda18fd85d4a4eb893022d6bfad846b1",
          "debug_file":"file://foo.invalid/demo.wasm"
        }]"#,
    )
    .unwrap();

    let stacktraces = serde_json::from_str(
        r#"[{
          "frames":[{
            "instruction_addr":"0x8c",
            "addr_mode":"rel:0"
          }]
        }]"#,
    )
    .unwrap();

    let request = make_symbolication_request(modules, stacktraces, vec![source]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_source_candidates() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_symsrv, source) = test::symbol_server();

    // its not wasm, but that is the easiest to write tests because of relative
    // addressing ;-)
    let modules = serde_json::from_str(
        r#"[{
          "type":"wasm",
          "debug_id":"7f883fcd-c553-36d0-a809-b0150f09500b",
          "code_id":"7f883fcdc55336d0a809b0150f09500b"
        }]"#,
    )
    .unwrap();

    let stacktraces = serde_json::from_str(
        r#"[{
          "frames":[{
            "instruction_addr":"0x3880",
            "addr_mode":"rel:0"
          }]
        }]"#,
    )
    .unwrap();

    let request = make_symbolication_request(modules, stacktraces, vec![source]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_dotnet_integration() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = test::symbol_server();

    let modules = serde_json::from_str(
        r#"[{
          "type":"pe_dotnet",
          "debug_file":"integration.pdb",
          "debug_id":"0c1033f78632492e91c6c314b72e1920e60b819d"
        }]"#,
    )
    .unwrap();

    let stacktraces = serde_json::from_str(
        r#"[{
          "frames":[{
            "instruction_addr": 10,
            "function_id": 6,
            "addr_mode":"rel:0"
          },
          {
            "instruction_addr": 6,
            "function_id": 5,
            "addr_mode": "rel:0"
          },
          {
            "instruction_addr": 0,
            "function_id": 3,
            "addr_mode": "rel:0"
          },
          {
            "instruction_addr": 0,
            "function_id": 2,
            "addr_mode": "rel:0"
          },
          {
            "instruction_addr": 45,
            "function_id": 1,
            "addr_mode": "rel:0"
          }]
        }]"#,
    )
    .unwrap();

    let request = make_symbolication_request(modules, stacktraces, vec![source]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_dotnet_embedded_sources() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = test::symbol_server();

    let modules = serde_json::from_str(
        r#"[{
          "type":"pe_dotnet",
          "debug_file":"portable-embedded.pdb",
          "debug_id":"b6919861-510c-4887-9994-943f64f70c37-870b9ef9"
        }]"#,
    )
    .unwrap();

    let stacktraces = serde_json::from_str(
        r#"[{
          "frames":[{
            "instruction_addr": 47,
            "function_id": 5,
            "addr_mode":"rel:0"
          }]
        }]"#,
    )
    .unwrap();

    let request = make_symbolication_request(modules, stacktraces, vec![source]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());
}

#[tokio::test]
async fn test_nuget_file() {
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let source = test::nuget_source();

    let modules = serde_json::from_str(
        r#"[{
          "type":"pe_dotnet",
          "code_id": "efc9a199e000",
          "code_file": "./TimeZoneConverter.dll",
          "debug_id": "4e2ca887-825e-46f3-968f-25b41ae1b5f3-9e6d3fcc",
          "debug_file": "./TimeZoneConverter.pdb",
          "debug_checksum": "SHA256:87a82c4e5e82f386968f25b41ae1b5f3cc3f6d9e79cfb4464f8240400fc47dcd"
        }]"#,
    )
    .unwrap();

    let stacktraces = serde_json::from_str(
        r#"[{
          "frames":[{
            "instruction_addr": "0x21",
            "function_id": "0xc",
            "addr_mode":"rel:0"
          }]
        }]"#,
    )
    .unwrap();

    let request = make_symbolication_request(modules, stacktraces, vec![source]);
    let response = symbolication.symbolicate(request).await;

    assert_snapshot!(response.unwrap());
}
