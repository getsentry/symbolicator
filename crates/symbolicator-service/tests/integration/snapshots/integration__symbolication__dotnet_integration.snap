---
source: crates/symbolicator-service/tests/integration/symbolication.rs
assertion_line: 277
expression: response.unwrap()
---
stacktraces:
  - frames:
      - status: symbolicated
        original_index: 0
        addr_mode: "rel:0"
        instruction_addr: "0xa"
        function_id: "0x6"
        lang: csharp
        filename: Program.cs
        abs_path: /Users/swatinem/Coding/sentry-dotnet/samples/foo/Program.cs
        lineno: 55
      - status: symbolicated
        original_index: 1
        addr_mode: "rel:0"
        instruction_addr: "0x6"
        function_id: "0x5"
        lang: csharp
        filename: Program.cs
        abs_path: /Users/swatinem/Coding/sentry-dotnet/samples/foo/Program.cs
        lineno: 48
      - status: symbolicated
        original_index: 2
        addr_mode: "rel:0"
        instruction_addr: "0x0"
        function_id: "0x3"
        lang: csharp
        filename: Program.cs
        abs_path: /Users/swatinem/Coding/sentry-dotnet/samples/foo/Program.cs
        lineno: 41
      - status: symbolicated
        original_index: 3
        addr_mode: "rel:0"
        instruction_addr: "0x0"
        function_id: "0x2"
        lang: csharp
        filename: Program.cs
        abs_path: /Users/swatinem/Coding/sentry-dotnet/samples/foo/Program.cs
        lineno: 36
      - status: symbolicated
        original_index: 4
        addr_mode: "rel:0"
        instruction_addr: "0x2d"
        function_id: "0x1"
        lang: csharp
        filename: Program.cs
        abs_path: /Users/swatinem/Coding/sentry-dotnet/samples/foo/Program.cs
        lineno: 18
modules:
  - debug_status: found
    features:
      has_debug_info: true
      has_unwind_info: false
      has_symbols: false
      has_sources: false
    arch: unknown
    type: pe_dotnet
    debug_id: 0c1033f7-8632-492e-91c6-c314b72e1920-e60b819d
    debug_file: integration.pdb
    image_addr: "0x0"
    candidates:
      - source: local
        location: "http://localhost:<port>/download/integration.pdb/0C1033F78632492E91C6C314B72E1920e60b819d/integration.src.zip"
        download:
          status: notfound
      - source: local
        location: "http://localhost:<port>/download/integration.pdb/0C1033F78632492E91C6C314B72E1920ffffffff/integration.pdb"
        download:
          status: ok
          features:
            has_debug_info: true
            has_unwind_info: false
            has_symbols: false
            has_sources: false
        debug:
          status: ok
      - source: local
        location: "http://localhost:<port>/download/integration.pdb/0C1033F78632492E91C6C314B72E1920ffffffff/integration.src.zip"
        download:
          status: notfound

